pkg/model/bootstrapscript.go

```diff
@@ -158,7 +184,84 @@ func (b *BootstrapScript) getRelevantHooks(hooks []kops.HookSpec, role kops.Inst
 			}
 		}
 	}
-	return relevantHooks
+
+	hooks := []kops.HookSpec{}
+	if len(relevantHooks) > 0 {
+		for _, hook := range relevantHooks {
+			if hook.Manifest != "" {
+				manifestFingerprint, err := b.computeFingerprint(hook.Manifest)
+				if err != nil {
+					return nil, err
+				}
+				hook.Manifest = manifestFingerprint + " (fingerprint)"
+			}
+
+			if hook.ExecContainer != nil && hook.ExecContainer.Command != nil {
+				execContainerCommandFingerprint, err := b.computeFingerprint(hook.ExecContainer.Command)
+				if err != nil {
+					return nil, err
+				}
+				hook.ExecContainer.Command = []string{execContainerCommandFingerprint + " (fingerprint)"}
+			}
+
+			hook.Roles = nil
+			hooks = append(hooks, hook)
+		}
+	}
+
+	return hooks, nil
+}
+
+// getRelevantFileAssets returns a list of file assets to be applied to the
+// instance group, with the Content fingerprinted to reduce size
+func (b *BootstrapScript) getRelevantFileAssets(allFileAssets []kops.FileAssetSpec, role kops.InstanceGroupRole) ([]kops.FileAssetSpec, error) {
+	relevantFileAssets := []kops.FileAssetSpec{}
+	for _, fileAsset := range allFileAssets {
+		if len(fileAsset.Roles) == 0 {
+			relevantFileAssets = append(relevantFileAssets, fileAsset)
+			continue
+		}
+		for _, fileAssetRole := range fileAsset.Roles {
+			if role == fileAssetRole {
+				relevantFileAssets = append(relevantFileAssets, fileAsset)
+				break
+			}
+		}
+	}
+
+	fileAssets := []kops.FileAssetSpec{}
+	if len(relevantFileAssets) > 0 {
+		for _, fileAsset := range relevantFileAssets {
+			if fileAsset.Content != "" {
+				contentFingerprint, err := b.computeFingerprint(fileAsset.Content)
+				if err != nil {
+					return nil, err
+				}
+				fileAsset.Content = contentFingerprint + " (fingerprint)"
+			}
+
+			fileAsset.Roles = nil
+			fileAssets = append(fileAssets, fileAsset)
+		}
+	}
+
+	return fileAssets, nil
+}
+
+// computeFingerprint takes an object and returns a base64 encoded fingerprint
+func (b *BootstrapScript) computeFingerprint(obj interface{}) (string, error) {
+	hasher := sha1.New()
+
+	data, err := kops.ToRawYaml(obj)
+	if err != nil {
+		return "", err
+	}
+
+	if _, err := hasher.Write(data); err != nil {
+		return "", fmt.Errorf("error computing fingerprint hash: %v", err)
+	}
+
+	return base64.StdEncoding.EncodeToString(hasher.Sum(nil)), nil
```

Comment:
I like this fingerprint function - it is unique enough that we could use it to retrieve the Content using only the ContentHash (within the "safe space" of a cluster!)

Reply:
I think you wrote this one in an earlier PR, came in quite handy :D